File: src/me/mars/triangles/ui/ImageGrid.java
package me.mars.triangles.ui;

import arc.Core;
import arc.graphics.Color;
import arc.graphics.Pixmap;
import arc.graphics.Texture;
import arc.graphics.g2d.*;
import arc.input.KeyCode;
import arc.math.Mathf;
import arc.math.geom.Point2;
import arc.math.geom.Vec2;
import arc.scene.Element;
import arc.scene.event.ClickListener;
import arc.scene.event.InputEvent;
import arc.scene.ui.layout.Scl;
import arc.util.Log;
import arc.util.Nullable;
import arc.util.Strings;
import arc.util.pooling.Pools;
import me.mars.triangles.SchemBuilder;
import mindustry.Vars;
import mindustry.ui.Fonts;

import static me.mars.triangles.PicToTri.debugMode;

public class ImageGrid extends Element {
	private @Nullable TextureRegion region;
	int sx = 0, sy = 0;
	float panX = 0f, panY = 0f, zoom = 1f;

	ConverterDialog dialog;

	public ImageGrid(ConverterDialog dialog) {
		this.dialog = dialog;
		this.addListener(new DragListener(KeyCode.mouseRight) {
			@Override
			public void dragged(float xDelta, float yDelta) {
				float scaledSize = scaled(ImageGrid.this.zoom);
				xDelta/= scaledSize/2f;
				yDelta/= scaledSize/2f;
				ImageGrid.this.panX+= xDelta;
				ImageGrid.this.panY+= yDelta;
				ImageGrid.this.clampPos();
			}

			@Override
			public void enter(InputEvent event, float x, float y, int pointer, Element fromActor) {
				ImageGrid.this.requestScroll();
				// REMOVEME
				ImageGrid.this.requestKeyboard();
			}

			@Override
			public boolean scrolled(InputEvent event, float x, float y, float amountX, float amountY) {
				ImageGrid.this.zoom = Mathf.clamp(ImageGrid.this.zoom - amountY/10f, 0.1f, 10f);
				ImageGrid.this.clampPos();
				return true;
			}
		});
		this.addListener(new ClickListener(){
			// REMOVEME
			@Override
			public boolean keyTyped(InputEvent event, char character) {
				switch (character) {
					case 'r' -> {
						ImageGrid.this.panX = ImageGrid.this.panY = 0f;
						return true;
					}
					case 'u' -> {
						ImageGrid.this.panY++;
						return true;
					}
					default -> {
						return false;
					}
				}
			}

			@Override
			public void clicked(InputEvent event, float x, float y) {
				if (event.keyCode != KeyCode.mouseLeft) return;
				float scl = scaled(ImageGrid.this.zoom);
				x-= ImageGrid.this.width/2f;
				y-= ImageGrid.this.height/2f;
				x-= ImageGrid.this.panX * scl/2f;
				y-= ImageGrid.this.panY * scl/2f;
				// Remove the logic range padding
				int sx = (int) (x/scl - dialog.lBlock.range/Vars.tilesize);
				int sy = (int) (y/scl - dialog.lBlock.range/Vars.tilesize);
				sx/= dialog.lDisplay.size;
				sy/= dialog.lDisplay.size;
//				Log.info("select [@, @] at @, @", sx, sy);
				if (sx < 0 || sx >= dialog.xChunks || sy < 0 || sy >= dialog.yChunks) return;
				ImageGrid.this.sx = sx;
				ImageGrid.this.sy = sy;
				dialog.select(sx, sy);
			}
		});
	}

	public void setDrawable(@Nullable Pixmap pixmap) {
		if (this.region != null) {
			this.region.texture.dispose();
		}
		if (pixmap != null) {
			this.region = new TextureRegion(new Texture(pixmap));
		} else {
			this.region = null;
		}
	}

	@Override
	public void draw() {
		super.draw();
		Lines.rect(this.x, this.y, this.width, this.height);
		// Pre-drawing stuff
		if (this.region == null || !this.clipBegin()) return;
		float scaledSize = scaled(this.zoom);
		float ox = this.x+this.width/2f + this.panX*scaledSize/2f;
		float oy = this.y+this.height/2f + this.panY*scaledSize/2f;
		// pad = logic range * scaled size
		// ox <-> (scaledSize * builder width)

		Lines.rect(ox, oy, scaledSize * dialog.filler.width, scaledSize * dialog.filler.height);
		Fill.rect(this.panX + this.width/2f, this.panY + this.width/2f, scaledSize, scaledSize);
		// Drawing logic block positions
		for (SchemBuilder.Display display : dialog.filler.displays) {
			Draw.color(display.color);
			for (int i = 0; i < display.points.size; i++) {
				int pos = display.points.items[i];
				int x = Point2.x(pos);
				int y = Point2.y(pos);
				Fill.crect(ox + x*scaledSize, oy + y*scaledSize, scaledSize, scaledSize);
			}
		}
		Draw.color();
		// Draw the image
		int displayRes = (int) (dialog.lDisplay.size * scaledSize);
		int drawPad = (int) (dialog.lBlock.range / Vars.tilesize * scaledSize);
		float maxWidth = dialog.xChunks * displayRes;
		float maxHeight = dialog.yChunks * displayRes;
		float scl = Math.min(maxWidth / this.region.width, maxHeight / this.region.height);
		float iw = this.region.width * scl, ih = this.region.height * scl;
		Draw.rect(this.region, drawPad + ox + iw/2f, drawPad + oy + ih/2f, iw, ih);
		Lines.stroke(1f);
		for (int x = 0; x < dialog.xChunks; x++) {
			for (int y = 0; y < dialog.yChunks; y++) {
				Lines.rect(drawPad + ox + x*displayRes, drawPad + oy + y*displayRes, displayRes, displayRes);
			}
		}
		this.clipEnd();
		// Draw coords
		StringBuilder builder = new StringBuilder();
		builder.append((int) this.panX).append(",").append((int) this.panY);
		builder.append(" : ").append(Strings.fixed(this.zoom, 1));
		if (this.hasMouse()) {
			Vec2 mouse = this.screenToLocalCoordinates(Core.input.mouse());
			mouse.sub(this.width/2f, this.height/2f);
			if (debugMode) builder.append(" Raw: ").append((int) mouse.x).append(",").append((int) mouse.y);
			mouse.sub(this.panX*scaledSize/2f, this.panY*scaledSize/2f);
			// TODO: Figure out why this is already scaled down to block size
			mouse.scl(1f/(scaledSize));
			builder.append(" @ ").append((int) mouse.x).append(",").append((int) mouse.y);
		}
		Font font = Fonts.outline;
		font.setColor(Color.white);
		font.getData().setScale(Scl.scl(1f));
		GlyphLayout layout = Pools.obtain(GlyphLayout.class, GlyphLayout::new);
		layout.setText(font, builder.toString());
		Draw.color(0, 0, 0, 0.5f);
		Fill.rect(this.x + layout.width/2f, this.y + layout.height/2f, layout.width, layout.height);
		Draw.color();
		font.draw(layout, this.x, this.y + layout.height/2f);
		Pools.free(layout);
	}

	public void clampPos() {
		float scaledSize = scaled(this.zoom);
		float xbounds = (this.width) / (scaledSize);
		float ybounds = (this.height) / (scaledSize);
		float w = dialog.filler.width;
		float h = dialog.filler.height;
		this.panX = Mathf.clamp(this.panX, -xbounds-w, xbounds-w);
		this.panY = Mathf.clamp(this.panY, -ybounds-h, ybounds-h);
	}

	private static float scaled(float scl) {
		return scl * Vars.tilesize * 0.5f;
	}
}

File: renderer/src/main/kotlin/me/mars/triangles/renderer/elem/ImageGrid.kt
package me.mars.triangles.renderer.elem

import arc.Core
import arc.graphics.Color
import arc.graphics.Pixmap
import arc.graphics.Texture
import arc.graphics.g2d.*
import arc.input.KeyCode
import arc.math.Mathf
import arc.math.geom.Point2
import arc.scene.Element
import arc.scene.event.ElementGestureListener
import arc.scene.event.InputEvent
import arc.scene.event.InputListener
import arc.scene.ui.layout.Scl
import arc.util.Nullable
import arc.util.Strings
import arc.util.pooling.Pools
import me.mars.triangles.layout.Layout
import me.mars.triangles.layout.LogicDisplayLayout
import me.mars.triangles.utils.Prefs
import mindustry.Vars
import mindustry.content.Blocks
import mindustry.ui.Fonts
import mindustry.world.blocks.logic.LogicDisplay
import kotlin.math.abs

class ImageGrid : Element() {
    var layout: Layout<*> = LogicDisplayLayout(Blocks.logicDisplay as LogicDisplay, 1, 1)
    var onSelect: (Point2) -> Unit = { }

    private var region: TextureRegion = TextureRegion(Texture(0, 0))
    var panX: Float = 0f
    var panY: Float = 0f
    var zoom: Float = 1f

    init {
        this.addListener(object : DragListener(KeyCode.mouseLeft) {
            override fun dragged(xDelta: Float, yDelta: Float) {
                var xDelta = xDelta
                var yDelta = yDelta
                val scaledSize: Float = scaled(this@ImageGrid.zoom)
                xDelta /= scaledSize / 2f
                yDelta /= scaledSize / 2f
                this@ImageGrid.panX += xDelta
                this@ImageGrid.panY += yDelta
                this@ImageGrid.clampPos()
            }

            override fun clicked(x: Float, y: Float) {
                /**
                 *             mouse.sub(this.width / 2f, this.height / 2f)
                 *             if (PicToTri.debugMode) builder.append(" Raw: ").append(mouse.x.toInt()).append(",").append(mouse.y.toInt())
                 *             mouse.sub(this.panX * scaledSize / 2f, this.panY * scaledSize / 2f)
                 *             // TODO: Figure out why this is already scaled down to block size
                 *             mouse.scl(1f / (scaledSize))
                 */
                var x = x
                var y = y
                val scl: Float = scaled(this@ImageGrid.zoom)
                x -= this@ImageGrid.width / 2f
                y -= this@ImageGrid.height / 2f
                x -= this@ImageGrid.panX * scl / 2f
                y -= this@ImageGrid.panY * scl / 2f
                val sx = (x/scl).toInt()
                val sy = (y/scl).toInt()
                if (sx < 0 || sx >= layout.width || sy < 0 || sy >= layout.height) return
                onSelect(Point2(sx, sy))
            }

            override fun enter(event: InputEvent?, x: Float, y: Float, pointer: Int, fromActor: Element?) {
                this@ImageGrid.requestScroll()
                this@ImageGrid.requestKeyboard()
            }

            override fun scrolled(event: InputEvent?, x: Float, y: Float, amountX: Float, amountY: Float): Boolean {
                this@ImageGrid.zoom = Mathf.clamp(this@ImageGrid.zoom - amountY / 10f, 0.1f, 15f)
                this@ImageGrid.clampPos()
                return true
            }

            override fun keyTyped(event: InputEvent?, character: Char): Boolean {
                if (character == 'r') {
                    this@ImageGrid.panX = 0f
                    this@ImageGrid.panY = 0f
                    return true
                }
                return false
            }
        })
        if (Vars.mobile) {
            this.addListener(object : ElementGestureListener() {
                var lastZoom: Float = 0f
                override fun zoom(event: InputEvent?, initialDistance: Float, distance: Float) {
                    this@ImageGrid.zoom = Mathf.clamp(distance / initialDistance * this.lastZoom, 0.1f, 15f)
                    this@ImageGrid.clampPos()
                }

                override fun touchUp(event: InputEvent?, x: Float, y: Float, pointer: Int, button: KeyCode?) {
                    this.lastZoom = this@ImageGrid.zoom
                }
            })
        }
    }

    fun setDrawable(pixmap: Pixmap?) {
        this.region.texture.dispose()
        this.region = TextureRegion(Texture(pixmap))
    }

    override fun draw() {
        super.draw()
        Lines.rect(this.x, this.y, this.width, this.height)
        // Pre-drawing stuff
        if (!this.clipBegin()) return
        val scaledSize: Float = scaled(this.zoom)
        // TODO This should be layout origin position, either rename better or leave comment
        val ox = this.x + this.width / 2f + this.panX * scaledSize / 2f
        val oy = this.y + this.height / 2f + this.panY * scaledSize / 2f
        // Border for whole schematic
        Lines.rect(ox, oy, scaledSize * layout.width, scaledSize * layout.height)
        if (Prefs.debugMode) {
            Fill.rect(this.panX + this.width / 2f, this.panY + this.width / 2f, scaledSize, scaledSize)
//            Draw.color(Pal.darkerMetal)
//            for (x in 0..<dialog.filler.width) {
//                for (y in 0..<dialog.filler.height) {
//                    if (dialog.filler.occupied(x, y)) Fill.crect(
//                        ox + x * scaledSize,
//                        oy + y * scaledSize,
//                        scaledSize,
//                        scaledSize
//                    )
//                }
//            }
//            Draw.color()
        }
        Draw.color()
        // Draw the image
        val ix = layout.imageBounds.x * scaledSize
        val iy = layout.imageBounds.y * scaledSize
        val iw = layout.imageBounds.width * scaledSize
        val ih = layout.imageBounds.height * scaledSize
        // TODO + or - iw/2?
        Draw.rect(this.region, ox + ix + iw/2f, oy + iy + ih/2f, iw, ih)
        // Draw chunks
        Lines.stroke(1f)
        for (chunk in layout.chunks) {
            val chunkWidth = chunk.width * scaledSize
            val chunkHeight = chunk.height * scaledSize
            Lines.rect(ox + chunk.chunkX*scaledSize, oy + chunk.chunkY*scaledSize, chunkWidth, chunkHeight)
        }
        // Draw schematic blocks
        for (tile in layout.preview.tiles) {
            val region = tile.block.fullIcon
            val tileSize = tile.block.size * scaledSize
            Draw.rect(region, ox + tile.x*tileSize + tileSize/2f, oy + tile.y*tileSize + tileSize/2f, tileSize, tileSize)
        }
        this.clipEnd()
        // Draw coords
        val builder = StringBuilder()
        builder.append(this.panX.toInt()).append(",").append(this.panY.toInt())
        builder.append(" : ").append(Strings.fixed(this.zoom, 1))
        if (this.hasMouse()) {
            val mouse = this.screenToLocalCoordinates(Core.input.mouse())
            mouse.sub(this.width / 2f, this.height / 2f)
            if (Prefs.debugMode) builder.append(" Raw: ").append(mouse.x.toInt()).append(",").append(mouse.y.toInt())
            mouse.sub(this.panX * scaledSize / 2f, this.panY * scaledSize / 2f)
            // TODO: Figure out why this is already scaled down to block size
            mouse.scl(1f / (scaledSize))
            builder.append(" @ ").append(mouse.x.toInt()).append(",").append(mouse.y.toInt())
        }
        val font = Fonts.outline
        font.color = Color.white
        font.getData().setScale(Scl.scl(1f))
        val glyphLayout = Pools.obtain(GlyphLayout::class.java) { GlyphLayout() }
        glyphLayout.setText(font, builder.toString())
        Draw.color(0f, 0f, 0f, 0.5f)
        Fill.rect(this.x + glyphLayout.width / 2f, this.y + glyphLayout.height / 2f, glyphLayout.width, glyphLayout.height)
        Draw.color()
        font.draw(glyphLayout, this.x, this.y + glyphLayout.height)
        Pools.free(glyphLayout)
    }

    fun clampPos() {
        val scaledSize: Float = scaled(this.zoom)
        val xbounds = (this.width) / (scaledSize)
        val ybounds = (this.height) / (scaledSize)
        val w = layout.width
        val h = layout.height
        this.panX = Mathf.clamp(this.panX, -xbounds - w, xbounds - w)
        this.panY = Mathf.clamp(this.panY, -ybounds - h, ybounds - h)
    }
}

private fun scaled(scl: Float): Float {
    // Pixels per world unit at this zoom
    return scl * Vars.tilesize * 0.5f
}

abstract class DragListener(var tapKey: KeyCode?) : InputListener() {
    private var lastX = 0f
    private var lastY = 0f
    private var startX = 0f
    private var startY = 0f

    var tapSquare: Float = 14f
    var pressed: Boolean = false

    override fun touchDown(event: InputEvent?, x: Float, y: Float, pointer: Int, button: KeyCode?): Boolean {
        if (Core.app.isMobile && pointer != 0) return false
        this.pressed = button == this.tapKey
        if (this.pressed) {
            this.startX = x
            this.startY = y
        }
        //		Log.info("Touchdown: @, @", x,y);
        this.lastX = x
        this.lastY = y
        return true
    }

    override fun touchDragged(event: InputEvent?, x: Float, y: Float, pointer: Int) {
        if ((Core.app.isMobile && pointer != 0)) return
        this.pressed = this.pressed and this.inSquare(x, y)
        this.dragged(x - this.lastX, y - this.lastY)
        this.lastX = x
        this.lastY = y
    }

    override fun touchUp(event: InputEvent?, x: Float, y: Float, pointer: Int, button: KeyCode?) {
        if (this.pressed && button == this.tapKey && this.inSquare(x, y)) this.clicked(x, y)
    }

    private fun inSquare(x: Float, y: Float): Boolean {
        return abs(this.startX - x) <= this.tapSquare && abs(this.startY - y) <= this.tapSquare
    }

    abstract fun dragged(xDelta: Float, yDelta: Float)

    abstract fun clicked(x: Float, y: Float)
}
